[clang][CodeGen][MSVC] Match how MSVC returns vector types from member functions

henrybw opened the pull request:
The MSVC ABI usually returns vector types directly, but on x86 and x86-64, there seems to be a special case for C++ member functions, which return vector types indirectly (with the exception of member functions using the `__vectorcall` calling convention, which return vector types > 64 bits directly).

This is an ABI change and has the potential to cause backward compatibility issues with previous Clang releases.

Fixes #104.

henrybw committed:
[clang][CodeGen][MSVC] Match how MSVC returns vector types from member functions

The MSVC ABI usually returns vector types directly, but on x86 and x86-64, there
seems to be a special case for C++ member functions, which return vector types
indirectly (with the exception of member functions using the `__vectorcall`
calling convention, which return vector types > 64 bits directly).

This is an ABI change and has the potential to cause backward compatibility
issues with previous Clang releases.

Fixes #104.

henrybw commented:
cc @RKSimon @rnk @efriedma-quic

Fznamznon reviewed (commented):
Hi, thank you for the patch. I wonder if it can make clang ABI incompatible with its previous versions? Perhaps it makes sense to put the new behavior under `-fclang-abi-compat` flag.

henrybw commented:
> Hi, thank you for the patch. I wonder if it can make clang ABI incompatible with its previous versions? Perhaps it makes sense to put the new behavior under `-fclang-abi-compat` flag.

Thanks for the review! Good catch: indeed, this change will break ABI compatibility and should respect `-fclang-abi-compat`. I have updated the patch (and release notes) to add ABI version 21, and now only enable this behavior for newer versions of the ABI.

I wasn't sure if this warranted adding a test case to `clang/test/CodeGenCXX/clang-abi-compat.cpp`, but if that makes more sense than adding it to my new `clang/test/CodeGenCXX/microsoft-abi-vector-types.cpp` test, I'd be happy to move it.

rnk reviewed (commented):
I worry that the real invariant here is that instance methods must always pass `this` in the first argument register (`RCX`), and we've only covered a few of the cases that enable indirect return. Like, I wonder if we can hit this via member pointers.

Regardless, I don't want to expand scope, I'd rather approve the PR and handle it in a follow-up. Please update the test and we can land this.

henrybw commented:
> I worry that the real invariant here is that instance methods must always pass `this` in the first argument register (`RCX`), and we've only covered a few of the cases that enable indirect return. Like, I wonder if we can hit this via member pointers.
> 
> Regardless, I don't want to expand scope, I'd rather approve the PR and handle it in a follow-up. Please update the test and we can land this.

To follow up on this, I dug into this some more, and found a couple of concerning discoveries:

* On x86-64, MSVC does *not* return `__m64` vectors directly from vectorcall member functions: https://godbolt.org/z/aa3E918xY. This is not consistent with how MSVC returns `__m128`, `__m256`, and `__m512` vectors from vectorcall member functions. But, it means that my patch does not adhere to the MSVC ABI for `__m64`.
* On ARM64, MSVC returns 64-bit and 128-bit vectors directly from member functions, matching Clang. However, MSVC *also* returns 256-bit and 512-bit vectors directly, whereas Clang returns them indirectly: https://godbolt.org/z/YP6M34zvq, https://godbolt.org/z/bzadn9Yre

(Note: I'm not as familiar with ARM64 as I am with x86/x86-64, so if the ABI discrepancy there is actually user error on my part, then my apologies.)

Given the discrepancy between `__m64` and `__m128` on x86-64, I wonder if this has something to do with how MSVC handles types that can map to SSE registers (and is not specific to just vector types).

Anyway, given these findings, do we still think it's a good idea to land this PR? I'm wary of causing further churn in the Clang ABI, because my patch amounts to a partial fix, and I'd be happy to back this out in favor of a more complete fix. (This is my first time contributing to LLVM, so I'm not sure what the right call here is.)

rnk reviewed (commented):
I think it is probably worth handling the `__m64` case before landing. If we had landed, I wouldn't insist on additional ABI compat checks before fixing the bug. We're mostly focused on stable ABIs between releases.

rnk commented:
Thanks for researching the `__m64` case!

henrybw commented on clang/lib/CodeGen/MicrosoftCXXABI.cpp:
```diff
@@ -1168,15 +1168,23 @@ static bool isTrivialForMSVC(const CXXRecordDecl *RD, QualType Ty,
+    // On x86, MSVC seems to only return vector types indirectly from non-
+    // vectorcall C++ instance methods.
+    isIndirectReturn =
+        CGM.getTarget().getTriple().isX86() && FI.isInstanceMethod() &&
```
Done.

henrybw commented:
> Thanks for researching the `__m64` case!

Thanks for the reviews! I have updated the PR to handle the `__vectorcall` `__m64` case correctly, and added test cases for 64-bit vectors on x86, x86-64, and ARM64. I have also reworded the commit message and the release note slightly to reflect the updated changes.

Looks like I'll also need to rebase this before landing (my changes to the Clang ABI versions header have a merge conflict), but I wanted to get this latest iteration out for review first before rebasing, so I don't blow away the context of the existing review comments.

henrybw commented:
ping

henrybw commented:
@rnk Friendly ping for review

rnk reviewed (approved):
Thanks for working on this! I'm sorry for the delay, work got in the way.

henrybw committed:
Rebase and bump ABI compatibility version to Clang 22

henrybw commented:
Hi @rnk! This PR kinda fell off my radar last winter (holidays, etc.) and didn't wind up making it into Clang 22, but I'd like to land it in Clang 23 if possible.

I've pushed an update that rebases the patch and updates the `-fclang-abi-compat` check and release note from ABI version 21 to 22, but the core logic is still the same.

Is your LGTM still good? If so, could you merge this for me? (I don't have commit access.)

Thank you!
